packages/world-consumer/src/experimental/WorldConsumer.sol

```diff
@@ -0,0 +1,94 @@
+import { System } from "@latticexyz/world/src/System.sol";
+
+abstract contract WorldConsumer is System {
+  bytes14 public immutable namespace;
+
+  error WorldConsumer_RootNamespaceNotAllowed(address worldAddress);
+  error WorldConsumer_NamespaceAlreadyExists(address worldAddress, bytes14 namespace);
+  error WorldConsumer_NamespaceDoesNotExists(address worldAddress, bytes14 namespace);
+  error WorldConsumer_CallerHasNoNamespaceAccess(address worldAddress, bytes14 namespace, address caller);
+  error WorldConsumer_CallerIsNotWorld(address worldAddress, address caller);
+  error WorldConsumer_ValueNotAllowed(address worldAddress);
+
+  modifier onlyWorld() {
+    address world = _world();
+    if (world != msg.sender) {
+      revert WorldConsumer_CallerIsNotWorld(world, msg.sender);
+    }
+    _;
+  }
+
+  modifier onlyNamespace() {
+    address world = _world();
+    if (world != msg.sender) {
+      revert WorldConsumer_CallerIsNotWorld(world, msg.sender);
+    }
+
+    // We use WorldContextConsumer directly as we already know the world is the caller
+    address sender = WorldContextConsumer._msgSender();
+    if (!ResourceAccess.get(getNamespaceId(), sender)) {
+      revert WorldConsumer_CallerHasNoNamespaceAccess(world, namespace, sender);
+    }
+
+    _;
+  }
+
+  constructor(IBaseWorld _world, bytes14 _namespace, bool registerNamespace) {
+    address worldAddress = address(_world);
+    StoreSwitch.setStoreAddress(worldAddress);
+
+    if (_namespace == bytes14(0)) {
+      revert WorldConsumer_RootNamespaceNotAllowed(worldAddress);
+    }
+
+    namespace = _namespace;
+
+    ResourceId namespaceId = WorldResourceIdLib.encodeNamespace(_namespace);
+    bool namespaceExists = ResourceIds.getExists(namespaceId);
+
+    if (registerNamespace) {
+      if (namespaceExists) {
+        revert WorldConsumer_NamespaceAlreadyExists(worldAddress, _namespace);
+      }
+      _world.registerNamespace(namespaceId);
+    } else if (!namespaceExists) {
+      revert WorldConsumer_NamespaceDoesNotExists(worldAddress, _namespace);
+    }
+  }
+
+  function getNamespaceId() public view returns (ResourceId) {
+    return WorldResourceIdLib.encodeNamespace(namespace);
+  }
+
+  function _msgSender() public view virtual override returns (address sender) {
+    return _world() == msg.sender ? WorldContextConsumer._msgSender() : msg.sender;
+  }
+
+  function _msgValue() public view virtual override returns (uint256 value) {
+    address world = _world();
+    if (world != msg.sender) {
+      revert WorldConsumer_ValueNotAllowed(world);
+    }
+
+    return WorldContextConsumer._msgValue();
+  }
+
+  function _encodeTableId(bytes16 name) internal view returns (ResourceId) {
+    return WorldResourceIdLib.encode(RESOURCE_TABLE, namespace, name);
+  }
+
+  function _encodeOffchainTableId(bytes16 name) internal view returns (ResourceId) {
+    return WorldResourceIdLib.encode(RESOURCE_OFFCHAIN_TABLE, namespace, name);
+  }
```

Comment:
I don't know if I like the extra surface area to maintain with this. Seems simple enough to encode these as needed?

But I do think this points to a need to have more helpers in `WorldResourceIdLib` to derive various resource IDs based on an existing resource ID.

Reply:
done, will create an issue for better helpers